convert last components to use accelerator framework

edgargabriel opened the pull request:
This pull request combines two major aspects that are hard to separate in order to make the CI pass:

Part 1:
- extend the accelerator API by three new functions
- add a new constant for default stream
- minor bug fixes in the existing API specification and implementation
- update the create_event interface to take a flag indicating whether the event is intended to be used IPC or not
- implementation of the new IPC functions in the rocm and null component
- add stubs for new functions to cuda and ze components

Part 2:
 - convert rcache/gpusm, rcache/rgpusm, and btl/smcuda to use the accelerator framework
 - introduce a new mca parameter controlling whether we use the sync_memops attribute or IPC events to ensure buffers are synchronized
 
Still to do:
 - find a solution for the lazy initialization problem, since btl/smcuda will face the same challenge as coll/cuda from that perspective
 - btl/smcuda has some ifdefs regarding GDRcopy support. Did not worry about that for now

wenduwan reviewed (commented):
Reviewed mca/accelerator to my knowledge

edgargabriel commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -110,19 +112,28 @@ struct opal_accelerator_stream_t {
+OBJ_CLASS_DECLARATION(opal_accelerator_stream_t);
+
+/* Constant indicating the default/zero stream */
+#define MCA_ACCELERATOR_STREAM_DEFAULT (opal_accelerator_stream_t *)0x00000002
```
@wenduwan this is intended to be a constant that rocm/ze/cuda can recognize. I wanted to have a value that is clearly not a valid address, hopefully unique in the accelerator framework API spec, but different from zero/NULL

edgargabriel commented on opal/mca/accelerator/accelerator.h:
```diff
@@ -110,19 +112,28 @@ struct opal_accelerator_stream_t {
+    opal_object_t super;
     size_t size;
     uint8_t handle[IPC_MAX_HANDLE_SIZE];
+    void* dev_ptr;
```
after opening an IPC handle, the function returns a device pointer that can contains the address under which the remote memory location be can accessed. We need to store this address since it is required during the CloseMemhandle function as an argument for both CUDA and ROCM. 

```
CUresult cuIpcCloseMemHandle (Cudeviceptr dptr );
hipError_t hipIpcCloseMemHandle (void *devPtr);
```

The destructor interfaces only provide the opal_accelerator_ipc_handle_t structure that we 'destruct', hence the pointer has to be stored as part of the handle.

edgargabriel commented on opal/mca/accelerator/rocm/accelerator_rocm_module.c:
```diff
@@ -158,7 +166,8 @@ static int mca_accelerator_rocm_create_stream(int dev_id, opal_accelerator_strea
 static void mca_accelerator_rocm_stream_destruct(opal_accelerator_rocm_stream_t *stream)
 {
-    if (NULL != stream->base.stream) {
+    if (MCA_ACCELERATOR_STREAM_DEFAULT != (opal_accelerator_stream_t *)stream &&
+	NULL != stream->base.stream) {
```
yes, it will need the same change

wenduwan replied:
Could you please do it in this change in case we forget?

edgargabriel replied:
I added the changes. It might not have been super necessary in the case of cuda because the cuda component didn't check for stream->stream being NULL, but I added it anyway just to be safe. I also added the GET_STREAM logic to cuda component as well.

edgargabriel replied:
@hppritcha I was looking into the ze component and whether I can apply the same changes to this component, but I honestly do not know enough about level zero API, and e.g. how the concept of the default stream would be handled there. I am however more than happy to work with you to resolve this, please feel free to ping me for that.

hppritcha replied:
i will take a look at ways to do this in ZE, but will have to wait till next week.

edgargabriel commented on opal/mca/accelerator/rocm/accelerator_rocm_module.c:
```diff
@@ -265,6 +276,27 @@ static int mca_accelerator_rocm_query_event(int dev_id, opal_accelerator_event_t
+        return OPAL_ERR_BAD_PARAM;
+    }
+
+    hipError_t err = hipStreamWaitEvent(MCA_ACCELERATOR_STREAM_DEFAULT == stream ? 
+                                        0 : *((hipStream_t *)stream->stream),
+					*((hipEvent_t *)event->event), 0);
```
would a macro also be ok?

edgargabriel commented:
@wenduwan thank you for your review, I think I addressed all of the points that you raised, please let me know if you have any other questions/comments

wenduwan commented on opal/mca/accelerator/rocm/accelerator_rocm_module.c:
```diff
@@ -456,31 +488,153 @@ static int mca_accelerator_rocm_get_address_range(int dev_id, const void *ptr, v
+                                        (hipDeviceptr_t)dev_ptr);
+    if (hipSuccess != err) {
+        opal_output_verbose(10, opal_accelerator_base_framework.framework_output,
+                            " Error in hipIpcGetMemHandle dev_ptr %p", dev_ptr);
```
nit extra space in print `_Error in ...`

wenduwan commented:
@edgargabriel YW I only have minor comments. I can test the PR in our CI if you think it's ready.

edgargabriel commented:
not entirely sure why the NVIDIA CI is failing.  I confirmed that the cuda component can compile. If I clock on the 'Details' link and want to see what the build failure is, my browser tab freezes and I do not get to see the actual error message

edgargabriel commented:
Never mind, I think I see the issue:

```
 make[2]: Entering directory '/GIT/ompi-640-1/opal/include'
  make[2]: *** No rule to make target 'opal/opal_cuda.h', needed by 'all-am'.  Stop.
  make[2]: Leaving directory '/GIT/ompi-640-1/opal/include'
  make[1]: *** [Makefile:1955: install-recursive] Error 1
  make[1]: Leaving directory '/GIT/ompi-640-1/opal'
  make: *** [Makefile:1521: install-recursive] Error 1
```

opal/include/opal/opal_cuda.h doesn't exist anymore, its now opal_gpu.h. I probably forgot to update that somewhere. Will investigate

edgargabriel commented:
Some performance numbers for smcuda, comparing it to UCX and sm. This is on an MI250 system. We can clearly see that the IPC path is being used, the SYNC_MEMOPS path/solution is quite a bit faster than using IPC events. However, even the IPC events path is faster than using a host staging buffer with sm.

![smcuda-perf](https://github.com/open-mpi/ompi/assets/8238622/bd43705d-83cc-4f83-bae4-132c8088eb8c)

We might need to analyze whether the IPC handle caching works as expected.

wenduwan reviewed (approved):
The PR passed our internal CI.

edgargabriel commented:
@wenduwan thank you. I will add the change to the smcuda priority that we discussed in slack.

The one additional item that I would like to make sure before merging the commit is that we do not break the ze component compilation, but I do not have the setup to test this. @hppritcha not sure whether there is an easy way to create at github workflow step for the ze component, similar to what we do for rocm, to ensure that we at least cover compilation of the component.

hppritcha commented:
i will try to get to this today.

edgargabriel commented:
@hppritcha I just realized that you did some tests and added another commit to this pr ( I somehow completely missed it, not sure why github didn't notify me). Does this mean that we are good to merge this pull request, i.e. accelerator/ze still compiles and works?

hppritcha commented:
yes its ready to go as far as I can tell.
